File: 4771번： 아름다운 삼각형/아름다운 삼각형.cpp
#include <bits/stdc++.h>
using namespace std;

const double EPS_AREA = 1e-6;
const double EPS_SIDE = 1e-3;

// Function to compare two doubles with a given epsilon
bool isEqual(double a, double b, double eps) {
    return fabs(a - b) < eps;
}

// Function to compute distance between two points
double distanceBetween(double x1, double y1, double x2, double y2){
    double dx = x1 - x2;
    double dy = y1 - y2;
    return sqrt(dx*dx + dy*dy);
}

int main(){
    ios::sync_with_stdio(false);
    cin.tie(0);
    
    double x1, y1, x2, y2, x3, y3;
    
    while(cin >> x1){
        if(isEqual(x1, -1.0, EPS_AREA)){
            break;
        }
        // Read the remaining five tokens
        if(!(cin >> y1 >> x2 >> y2 >> x3 >> y3)){
            // Incomplete input, exit
            break;
        }
        
        // Compute side lengths
        double a = distanceBetween(x2, y2, x3, y3);
        double b = distanceBetween(x1, y1, x3, y3);
        double c = distanceBetween(x1, y1, x2, y2);
        
        // Compute area using cross product
        double area = fabs( (x2 - x1)*(y3 - y1) - (x3 - x1)*(y2 - y1) ) / 2.0;
        
        if(area <= EPS_AREA){
            cout << "Not a Triangle\n";
            continue;
        }
        
        // Classify by sides
        string sideClass;
        bool ab = isEqual(a, b, EPS_SIDE);
        bool bc = isEqual(b, c, EPS_SIDE);
        bool ac = isEqual(a, c, EPS_SIDE);
        
        if(ab && bc){
            sideClass = "Equilateral";
        }
        else if(ab || bc || ac){
            sideClass = "Isosceles";
        }
        else{
            sideClass = "Scalene";
        }
        
        // Classify by angles
        string angleClass;
        if(sideClass == "Equilateral"){
            angleClass = "Acute";
        }
        else{
            // Sort sides in ascending order
            vector<double> sides = {a, b, c};
            sort(sides.begin(), sides.end());
            double s1 = sides[0];
            double s2 = sides[1];
            double s3 = sides[2];
            
            // Compute squares
            double s1_sq = s1 * s1;
            double s2_sq = s2 * s2;
            double s3_sq = s3 * s3;
            
            // Compare using epsilon
            if(isEqual(s1_sq + s2_sq, s3_sq, 1e-3)){
                angleClass = "Right";
            }
            else if((s1_sq + s2_sq) > s3_sq + 1e-3){
                angleClass = "Acute";
            }
            else{
                angleClass = "Obtuse";
            }
        }
        
        // Output the classification
        cout << sideClass << " " << angleClass << "\n";
    }
}

File: 4771번: 아름다운 삼각형/아름다운 삼각형.py
import sys
import math

def is_triangle(p1, p2, p3):
    area = abs(
        (p1[0]*(p2[1]-p3[1]) +
         p2[0]*(p3[1]-p1[1]) +
         p3[0]*(p1[1]-p2[1])) / 2.0)
    return area > 1e-8

def distance(p1, p2):
    return math.hypot(p1[0] - p2[0], p1[1] - p2[1])

def classify_sides(a, b, c):
    eps = 1e-6
    if abs(a - b) < eps and abs(b - c) < eps:
        return "Equilateral"
    elif abs(a - b) < eps or abs(b - c) < eps or abs(c - a) < eps:
        return "Isosceles"
    else:
        return "Scalene"
    
def classify_angles(a, b, c):
    angles = []

    angle_A = math.acos((b ** 2 + c ** 2 - a ** 2) / (2 * b * c)) * (180 / math.pi)
    angle_B = math.acos((a ** 2 + c ** 2 - b ** 2) / (2 * a * c)) * (180 / math.pi)
    angle_C = 180 - angle_A - angle_B
    angles.extend([angle_A, angle_B, angle_C])

    max_angle = max(angles)
    if 88 <= max_angle <= 92:
        return "Right"
    elif max_angle > 90:
        return "Obtuse"
    else:
        return "Acute"

lines = sys.stdin.read().splitlines()
i = 0
while i < len(lines):
    line = lines[i]
    if line.strip() == '-1':
        break
    else:
        data = []
        while len(data) < 6:
            tokens = line.strip().split()
            data.extend(map(float, tokens))
            i += 1
            if len(data) < 6 and i < len(lines):
                line = lines[i]
            elif len(data) < 6:
                break
        if len(data) < 6:
            break
        x1, y1, x2, y2, x3, y3 = data
        p1 = (x1, y1)
        p2 = (x2, y2)
        p3 = (x3, y3)

        if not is_triangle(p1, p2, p3):
            print("Not a Triangle")
        else:
            a = distance(p2, p3)
            b = distance(p1, p3)
            c = distance(p1, p2)

            side_class = classify_sides(a, b, c)
            angle_class = classify_angles(a, b, c)
            print(f"{side_class} {angle_class}")
    if i >= len(lines):
        break